docs/software/ospool-containers.md

```diff
@@ -0,0 +1,91 @@
+which could come from a local install on each Execution Point or from CVMFS.
+If detected, then `HAS_SINGULARITY` is set to `true` and
+_all_ user payloads will be run in containers.
+(It is possible to override this behavior through site-specific pilot hackery.)
+
+For Backfill Containers, the container itself includes an installation of Apptainer (née Singularity)
+and it will always be used (in unprivileged mode) instead of any system or other installation.
+
+Today, a pilot runs the “user job wrapper” script,
+
+The user job is a vanilla job.
+Today, though, a pilot runs the “user job wrapper” script,
+which is a replacement for the user’s executable that does some stuff and
+then runs the user’s executable.
+If the `HAS_SINGULARITY` attribute is `true` in the environment of the wrapper,
+then it runs the actual user payload job in Singularity (or Apptainer).
+
+## PID Namespaces
+
+PID namespaces are a key technology that enables containers to isolate from each other.
+See, for example,
+[this Ubuntu copy of the man page](https://manpages.ubuntu.com/manpages/bionic/man7/pid_namespaces.7.html)
+for `pid_namespaces`.
+The `root` user always has the ability to create PID namespaces, so a privileged container runtime
+(i.e., not unprivileged Singularity) can always do this.
+
+There is a special feature for _user PID namespaces_, which can be created by unprivileged processes.
+To work, there is a certain kernel setting that must be set,
+and then user PID namespaces are available to all.
+
+Today, when the GlideinWMS user job wrapper is about to start a user payload job in a Singularity container,
+the default is to pass a flag to Singularity to use user PID namespaces.
+However, if user PID namespaces are not available (say, due to the kernel setting),
+the pilot start-up scripts do not detect this condition automatically.
+So, there is a special envirnoment variable that can be set on the outer container
+(e.g., the Backfill Container) to disable this flag:
+
+```SINGULARITY_DISABLE_PID_NAMESPACES=1```
+
+While set on the outer container, it affects only the user job wrapper and
+how it invokes the inner Singularity container.
+
+Mat says that HTCondor 10.2.2 has a feature to detect the lack of user PID namespaces and,
+in such a case, to avoid using them, but it is not clear that that feature will help anything
+as long as the GlideinWMS user job wrapper is being used.
+
+## Containers and GPUs
+
+Most, but certainly not all, OSPool payload jobs that request GPUs also request to run within a container.
+This is probably due to GPU-using software stacks often being complex,
+and so users often turn to containers for their runtime environment,
+including ones that we and others pre-build.
+
+Specifically, the Backfill Container images do not include the NVIDIA CUDA drivers
+(although, we are experimenting with that),
+so a site that contributes capacity using Backfill Containers
+and that wants to run GPU jobs _must_ run them under Apptainer.
```

Comment:
I was somewhat wrong about this (per the discussion in Slack earlier today). We do not include the CUDA drivers but instead we mount them in from outside (so we use the CUDA drivers the admin installed on the system outside the pilot).  This works everywhere but Expanse (a phrase you will hear often); the experimental images are for getting GPUs working for Expanse.

Hence it is _not_ true that GPU jobs _must_ be run under Apptainer, though your previous paragraph still applies...

Reply:
Ok, I made some changes to try to address this point.